parser/prism/Factory.cc

```diff
@@ -0,0 +1,512 @@
+                               .elements = elemNodes,
+                               .opening_loc = parser.convertLocOffsets(loc.copyWithZeroLength()),
+                               .closing_loc = parser.convertLocOffsets(loc.copyEndWithZeroLength())};
+
+    return up_cast(array);
+}
+
+pm_node_t *Factory::Block(core::LocOffsets loc, pm_node_t *body) const {
+    pm_block_node_t *block = allocateNode<pm_block_node_t>();
+
+    pm_location_t zeroLoc = parser.getZeroWidthLocation();
+
+    *block = (pm_block_node_t){.base = initializeBaseNode(PM_BLOCK_NODE, parser.convertLocOffsets(loc)),
+                               .locals = {.size = 0, .capacity = 0, .ids = nullptr},
+                               .parameters = nullptr,
+                               .body = body,
+                               .opening_loc = zeroLoc,
+                               .closing_loc = zeroLoc};
+
+    return up_cast(block);
+}
+
+pm_node_t *Factory::T_Array(core::LocOffsets loc) const {
+    return ConstantPathNode(loc, T(loc), "Array"sv);
+}
+
+pm_node_t *Factory::T_Class(core::LocOffsets loc) const {
+    return ConstantPathNode(loc, T(loc), "Class"sv);
+}
+
+pm_node_t *Factory::T_Enumerable(core::LocOffsets loc) const {
+    return ConstantPathNode(loc, T(loc), "Enumerable"sv);
+}
+
+pm_node_t *Factory::T_Enumerator(core::LocOffsets loc) const {
+    return ConstantPathNode(loc, T(loc), "Enumerator"sv);
+}
+
+pm_node_t *Factory::T_Hash(core::LocOffsets loc) const {
+    return ConstantPathNode(loc, T(loc), "Hash"sv);
+}
+
+pm_node_t *Factory::T_Set(core::LocOffsets loc) const {
+    return ConstantPathNode(loc, T(loc), "Set"sv);
+}
+
+pm_node_t *Factory::T_Range(core::LocOffsets loc) const {
+    return ConstantPathNode(loc, T(loc), "Range"sv);
+}
+
+pm_node_t *Factory::StatementsNode(core::LocOffsets loc, const vector<pm_node_t *> &body) const {
+    pm_statements_node_t *stmts = allocateNode<pm_statements_node_t>();
+    *stmts = (pm_statements_node_t){.base = initializeBaseNode(PM_STATEMENTS_NODE, parser.convertLocOffsets(loc)),
+                                    .body = {.size = 0, .capacity = 0, .nodes = nullptr}};
+
+    for (auto *node : body) {
+        pm_node_list_append(&stmts->body, node);
+    }
+
+    return up_cast(stmts);
+}
+
+void *Factory::malloc(size_t size) const {
+    void *p = ::xmalloc(size); // see Prism's `include/prism/defines.h`
+    if (!p) {
+        throw std::bad_alloc{};
+    };
+    return p;
+}
+
+void *Factory::calloc(size_t count, size_t size) const {
+    void *p = ::xcalloc(count, size); // see Prism's `include/prism/defines.h`
+    if (!p) {
+        throw std::bad_alloc{};
+    };
+    return p;
+}
+
+void *Factory::realloc(void *ptr, size_t size) const {
+    void *p = ::xrealloc(ptr, size); // see Prism's `include/prism/defines.h`
+    if (!p) {
+        throw std::bad_alloc{};
+    };
+    return p;
+}
+
+void Factory::free(void *ptr) const { // see Prism's `include/prism/defines.h`
+    ENFORCE(ptr);
```

Comment:
Does Prism's `xfree` really require this?  It looks like Prism defaults to using `free`, and `free` is certainly OK with null pointers.